composer/distributed/fsdp2_utils.py

```diff
@@ -260,30 +303,78 @@ def generate_default_policy(parent_model: nn.Module) -> CustomPolicy:
-            # Unless the user specifically sets the _fsdp_wrap attribute to False for the parent model,
-            # we default to wrapping the parent model.
-            ret = True
-        return ret
+            return bool(current_module._fsdp_wrap)
+        # TODO: make this recursive for reusability, similar to meta_init in param_init.py
+        if hasattr(parent_model, 'fsdp_wrap_fn') and isinstance(parent_model.fsdp_wrap_fn, Callable):
+            res = parent_model.fsdp_wrap_fn(current_module)
+            # Ensure all keys in the returned dict are valid FSDP2Config attributes
+            if isinstance(res, dict) and not set(res.keys()).issubset(FSDP2Config.settable_attrs()):
+                raise KeyError(
+                    f'Invalid FSDP2 config keys in wrap_fn return value. Valid keys are: {FSDP2Config.settable_attrs()}',
+                )
+            return res
+        return False
+
+    return CustomPolicy(lambda_fn)
+
+
+def generate_composer_model_policy(composer_model: ComposerModel) -> CustomPolicy:
+    """Generates a FSDP wrap policy for ComposerModel that mimics FSDP1 behavior.
+
+    This policy wraps all direct children of the ComposerModel but not the ComposerModel itself,
+    which matches the behavior of FSDP1's prepare_fsdp_module function. It also respects
+    any _fsdp_wrap attributes or fsdp_wrap_fn functions defined on modules.
+
+    Args:
+        composer_model (ComposerModel): The ComposerModel to generate a policy for.
+
+    Returns:
+        CustomPolicy: A policy function that determines which modules to wrap with FSDP.
+
+    Raises:
+        KeyError: If a module's fsdp_wrap_fn returns a dict with invalid FSDP2Config keys.
+    """
+    # Filter the specific deprecation warning to only appear once.
+    warnings.filterwarnings(
+        'once',
+        category=DeprecationWarning,
+        message='The _fsdp_wrap attribute will be removed in a future release. Please use fsdp_wrap_fn instead.',
+    )
+
+    cached_submodules_to_wrap: dict[nn.Module, bool | dict[str, Any]] = {composer_model: False}
+    for child in composer_model.children():
+        if isinstance(child, Metric | MetricCollection):
+            for module in child.modules():
+                cached_submodules_to_wrap[module] = False
+            continue
+        # this can be overwritten by the _fsdp_wrap attribute or fsdp_wrap_fn
+        cached_submodules_to_wrap[child] = True
+        fsdp_wrap_fn = getattr(child, 'fsdp_wrap_fn', lambda x: cached_submodules_to_wrap.get(x, False))
+        for module in child.modules():
+            if hasattr(module, '_fsdp_wrap'):
+                warnings.warn(
+                    DeprecationWarning(
+                        'The _fsdp_wrap attribute will be removed in a future release. Please use fsdp_wrap_fn instead.',
+                    ),
+                )
+                cached_submodules_to_wrap[module] = bool(module._fsdp_wrap)
+            elif module == child:
```

Comment:
nit: would this be better off as "child_module" instead of "module" and "child_root" instead of "child"? I think it would make this a bit more readable. I'm not entirely opposed to keeping it as is though